SIR model implementation first step

Zavioer opened the pull request.

Zavioer committed:
SIR model implementation first step

amasend commented on simulation/groups_interface.py:
```diff
@@ -0,0 +1,120 @@
+    Abstract class for specific population groups based on SIR model.
+    """
+    @abstractmethod
+    def __init__(self, x: float, y: float, label: str, inf_dist: float,
```
What are these parameters? It will be helpful to see a docstring with these parameters above, where you describe entire class.

amasend commented on simulation/groups_interface.py:
```diff
@@ -0,0 +1,120 @@
+        self.death_probability = death_prob
+
+    @abstractmethod
+    def __del__(self):
```
Why do you want to define a destructor? This is probably taken from C++ right?
    In Python this is highly unwanted. As Python uses the garbage collector to manage
    a memory allocation during runtime, we should not take any manual actions here.
    Here is a nice article explaining what is really happening with objects that need to be destroyed in Python:
    https://eli.thegreenplace.net/2009/06/12/safely-using-destructors-in-python/

amasend commented on simulation/groups_interface.py:
```diff
@@ -0,0 +1,120 @@
+        member_x: float, required
+            Object to compare position with.
+        member_y: float, required
+
```
Some missing docstring part?

amasend commented on simulation/groups.py:
```diff
@@ -0,0 +1,342 @@
+
+        can_move = True
+
+        if not 0 < self.x + x_distance < box_width:
```
ok I can see a reverse logic here with `not` statement, it is ok but in my opinion harder to read

amasend commented on simulation/groups.py:
```diff
@@ -0,0 +1,342 @@
+            self.x += x_distance
+            self.y += y_distance
+
+            return True
```
Why you defined `can_move` variable if you do not use it at the end (maybe it is better to keep the logic and return this variable?).

amasend commented on simulation/groups.py:
```diff
@@ -0,0 +1,342 @@
+            return False
+
+    def is_in_area(self, member_x: float, member_y: float, member_inf_dis: float) -> bool:
+        if self.x > member_x:
```
To calculate an absolute distance you can use an `abs()` function built in Python.

amasend commented on simulation/groups.py:
```diff
@@ -0,0 +1,342 @@
+        else:
+            y_difference = member_y - self.y
+
+        if x_difference < member_inf_dis or \
```
I assume that infection distance is a radius right? So to calculate a true Cartesian distance between two points on a plane, you should rather use this formula: https://www.thoughtco.com/understanding-the-distance-formula-2312242

amasend commented on simulation/groups.py:
```diff
@@ -0,0 +1,342 @@
+        else:
+            return False
+
+    def is_in_area(self, member_x: float, member_y: float, member_inf_dis: float) -> bool:
```
Maybe better name for this method is: `is_in_infection_area`? Or `inside_infection_area`? Better to be more explicit than implicit.

amasend commented on simulation/groups.py:
```diff
@@ -0,0 +1,342 @@
+    death.__doc__ = groups_interface.PopulationGroup.death.__doc__
+
+
+s_member = SGroup(1, 1, "s_member1", 0.25, 0.25, 0.15, 0.10)
+i_member = IGroup(1, 1.25, "i_member1", 0.25, 0.25, 0.15, 0.10)
+
+if s_member.is_in_area(i_member.x, i_member.y, i_member.infection_distance):
+    print("{smem} is in infection area of {imem}".format(smem=s_member.label, imem=i_member.label))
+
+    i = 0
+
+    while not s_member.infect():
+        i += 1
+
+    print("{smem} is infected. New IGroup instance created.".format(smem=s_member.label))
+    print("Infected in {att} attempt".format(att=i))
+
+    i_member2 = IGroup(s_member.x, s_member.y, "i_member2", 0.25, 0.25, 0.15, 0.10)
+
+    del s_member
```
Why this is here? We should define some separate file to prepare tests.
Also, `del` statement should not be invoked, as Python has garbage collector. Even when you write `del something`, garbage collector will not remove this object from the memory, so this is dangerous to assume we already deleted something.

amasend commented on tests/test_interface.py:
```diff
@@ -0,0 +1,138 @@
+    def test_06__check_if_instance_died__return_bool(self):
+        self.assertIsInstance(self.r_member.death(), bool,
+                              msg="Method recover() was not return bool.")
+
+
+class DGroupTestCase(unittest.TestCase):
+    def setUp(self) -> None:
+        self.d_member = groups.DGroup(5, 5, "member1", 0.25, 0.3, 0.2, 0.2)
+        self.d_member2 = groups.DGroup(4, 2, "member2", 0.25, 0.3, 0.2, 0.2)
+
+    def test_01__check_if_constructor_work_correctly__return_SGroup_instance(self):
+        self.assertIsInstance(self.d_member, groups.DGroup,
+                              msg="Constructor was not return a SGroup instance.")
+
+    def test_02__check_if_change_instance_position_in_grid_system__return_bool(self):
+        self.assertIsInstance(self.d_member.move(1, 2, 10, 10), bool,
+                              msg="Method move() was not return a bool object.")
+
+    def test_03__check_if_method_correctly_is_in_area__return_bool(self):
+        self.assertIsInstance(self.d_member.is_in_area(self.d_member2.x,
+                                                       self.d_member2.y,
+                                                       self.d_member2.infection_distance), bool,
+                              msg="Method is_in_area() was not return "
+                                  "a bool object.")
+
+    def test_04__check_if_method_returns_new_instance__return_bool(self):
+        self.assertIsInstance(self.d_member.infect(), bool,
+                              msg="Method infect() was not return bool.")
+
+    def test_05__check_if_instance_was_recovered__return_bool(self):
+        self.assertIsInstance(self.d_member.recover(), bool,
+                              msg="Method recover() was not return bool.")
+
+    def test_06__check_if_instance_died__return_bool(self):
+        self.assertIsInstance(self.d_member.death(), bool,
+                              msg="Method recover() was not return bool.")
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Ok, good to have some unit tests, but do you think this is enough? Maybe we need to test a functionality? eg. what if ...
Maybe it will be better to write some end to end tests with full classes behavior? So lets define couple of use cases and check if classes will behave correctly, not only in the meaning of return type, but also in the meaning of correct coordinates update, infection changes etc.

amasend commented on simulation/groups.py:
```diff
@@ -0,0 +1,342 @@
+__all__ = [
```
'SGroup',
    'IGroup',
    'RGroup',
    'DGroup'

are the same in implementation? In that case why we even bother to separate them into different classes? Now I see a lot of code duplication.
Please consider if we can do it better. The main goal here is to have one class per one SIR behavior. Do S class instance can be dead? Or do I class instance can be infected once again? Or R class instance can be infected?
Right now I can see the same functionality per every class, in the end, I want to see different functionality per different class. If this is not clear, we can discuss it on Monday morning.

amasend reviewed (changes requested):
Some changes needed. Lets discuss this on Monday morning.

Zavioer committed:
SIR(D) classes improvement

Try to clean duplicated code and upgrade abstract class. Upgrade logic in method is_in_infection_area() and behaviour_probability(). Add 4 end to end test. Create configuration files for sphinx package.

Zavioer committed:
Merge branch 'dev' into simulation

Zavioer committed:
Delete introduction_and_learning

amasend commented on simulation/groups_interface.py:
```diff
@@ -4,117 +4,44 @@
     @abstractmethod
-    def __init__(self, x: float, y: float, label: str, inf_dist: float,
-                 infection_prob: float, recover_prob: float, death_prob: float) -> None:
+    def __init__(self, x: float, y: float, label: str, **kwargs) -> None:
```
Why we have **kwargs here?
This is implicit pass of parameters, it is always better to be explicit (better readability).
Anyway, I guess it is here because you wanted to keep the same parameters as in other class implementations.
But you could define different parameters for these methods without any issues. Child class could easily override parent method with own parameters, and this is ok.

amasend commented on simulation/groups_interface.py:
```diff
@@ -4,117 +4,44 @@
-
-    @abstractmethod
-    def infect(self):
+    def behaviour_probability(self, behaviour_prob: float) -> bool:
```
This method could be a `staticmethod` as you are not using any internal class variables here.

amasend commented on tests/test_interface.py:
```diff
@@ -4,135 +4,101 @@
 
```
Ok, I still do not see any simple test for groups behavior in one place aka. `container`.
But this could be a next task as you wish. So we should have your solution tested on some experiment.

amasend commented on simulation/groups.py:
```diff
@@ -6,337 +6,186 @@
+    ----------
+    infection_prob: float, required
+        Probability to infect it should be value between 0.0 and 1.0.
+    **kwargs:
```
I do not think this is the correct way of documenting parameters but I may be wrong.
From my perspective it is not an elegant way (still better to be explicit than implicit and pass / document all interface)

amasend reviewed (dismissed):
Ok, we can go further with this, I want to see if this implementation is useful in context of such experiment. So the next step is to define some interface / api for experiment and try to do some runs to verify if this code is correct (should behave similar to theoretical SIR model).

Zavioer committed:
Auto labeling implementation

Add auto labeling for S, I, R groups classes and modyfied docstrings. Set behaviour_probability() as staticmethod.

Zavioer committed:
Syntax update

Zavioer committed:
Update groups.py

amasend commented on simulation/groups.py:
```diff
@@ -6,337 +6,205 @@
 
-        if not 0 < self.x + x_distance < box_width:
-            can_move = False
+    id = count(1)
```
Be careful with class variables in Python (class variables are shared between class instances / objects). Here, every instance of RGroup will have this `id`.

amasend commented on simulation/groups_interface.py:
```diff
@@ -4,117 +4,47 @@
     Abstract class for specific population groups based on SIR model.
     """
+
+    id = count(0)
```
Can you explain why base class also defines id? Every child class will override this.

Zavioer replied:
I implement this auto labeling convention prefix_id but I am not sure if we need this in further visualisation or simple prefix will be enough? I thought that id should be defined in base class or constructor do not work correctly.

amasend commented on simulation/groups.py:
```diff
@@ -6,337 +6,205 @@
+    y: float, required
+        Instance starting position on Y axis on cartesian grid defined by
+        container wrapper.
+    prefix: str, optional
```
If something should not be changed, why we allow to change it?

amasend replied:
Maybe protected attributes will be handful here? Such as `_variable_name`.

Zavioer replied:
> If something should not be changed, why we allow to change it?

I want to give user a possibility to set own labeling prefix but I was not sure if it will cause troubles in visualisation implementation with diffrent labe names than default.

amasend reviewed (approved):
I can still approve this PR and want to see whole implementation and test with experiment, container and interactions between groups ;)
If something is not working just tell me, we can schedule some call to clarify what is wrong, do some brain storm what are the possibilities.

amasend commented:
Ohh in addition, maybe it will be better to split your tests into separate files? One file per one test class, it is very common ;) And one more thing, It is always worth to add some prints in tests, to generate test reports, but this is only a suggestion.
